File: app/src/main/java/com/tunjid/rcswitchcontrol/model/Payload.java
package com.tunjid.rcswitchcontrol.model;

import com.google.gson.Gson;

import java.io.Serializable;
import java.util.LinkedHashSet;
import java.util.Set;

/**
 * Payload class
 * <p>
 * Created by tj.dahunsi on 2/11/17.
 */

public class Payload implements Serializable {
    private static final Gson GSON = new Gson();

    private String key;
    private String data;
    private String action;
    private String response;
    private Set<String> commands = new LinkedHashSet<>();

    public String getKey() {
        return key;
    }

    public String getData() {
        return data;
    }

    public String getAction() {
        return action;
    }

    public String getResponse() {
        return response;
    }

    public Set<String> getCommands() {
        return commands;
    }

    public String serialize() {
        return GSON.toJson(this);
    }

    public static Payload deserialize(String input) {
        return GSON.fromJson(input, Payload.class);
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        Payload payload = new Payload();

        private Builder() {

        }

        public Builder setKey(String key) {
            payload.key = key;
            return this;
        }

        public Builder setData(String data) {
            payload.data = data;
            return this;
        }

        public Builder setAction(String action) {
            payload.action = action;
            return this;
        }

        public Builder setResponse(String response) {
            payload.response = response;
            return this;
        }

        public Builder addCommand(String command) {
            payload.commands.add(command);
            return this;
        }

        public Payload build() {
            return payload;
        }
    }

}

File: app/src/main/java/com/tunjid/rcswitchcontrol/model/Payload.kt
package com.tunjid.rcswitchcontrol.model

import com.google.gson.Gson
import java.io.Serializable
import java.util.*

/**
 * Payload class
 *
 *
 * Created by tj.dahunsi on 2/11/17.
 */

class Payload : Serializable {

    var key: String? = null
        private set
    var data: String? = null
        private set
    var action: String? = null
        private set
    var response: String? = null
        private set

    val commands = LinkedHashSet<String>()

    fun serialize(): String {
        return GSON.toJson(this)
    }

    class Builder {
        private var payload = Payload()

        fun setKey(key: String): Builder {
            payload.key = key
            return this
        }

        fun setData(data: String): Builder {
            payload.data = data
            return this
        }

        fun setAction(action: String): Builder {
            payload.action = action
            return this
        }

        fun setResponse(response: String): Builder {
            payload.response = response
            return this
        }

        fun addCommand(command: String): Builder {
            payload.commands.add(command)
            return this
        }

        fun build(): Payload {
            return payload
        }
    }

    companion object {
        private val GSON = Gson()

        fun deserialize(input: String): Payload {
            return GSON.fromJson(input, Payload::class.java)
        }

        fun builder(): Builder {
            return Builder()
        }
    }

}
